### DISCUSSION TITLE
How to create a N2 Diagram

### DISCUSSION THREAD

**Discussion by Mr-PDM** [OpenMDAO/OpenMDAO#3090] [General]
Hello guys, 

I'm working on a Optimization project for college where we are optimizing a horizontal stabilizer in an aircraft. I want to create de N2 Diagram for this project, but I'm a bit lost about how to do it, there are a lot of scripts and documents and I don't know where to start. I have defined the variables and the different connections between them, but I don't know how to write them so the scripts can recognize them or how to properly run the codes to obtain the N2 diagram.

Sorry, it is a vague question but  I'm not very familiar with this level of programming. 
Thanks for understanding and for your help, and good work because this repository is amazing :)

**Comment by JustinSGray** [comment]
Were you able to follow the [docs on creating a basic N2 diagram](https://openmdao.org/newdocs/versions/3.30.0/features/model_visualization/n2_basics/n2_basics.html)? If not, where did you get confused? 


The general workflow is to define a model, then use the openmdao command line tools to create an n2. You save your model script, than point the command line tools at it like: 

    openmdao n2 <your_run_script.py>


alternatively, you can generate the N2 directly in the script itself, by just passing your problem instance to the `om.n2()` function --- after you've called setup on the problem: 

```
p.setup()

om.n2(p)
```

**Reply by Mr-PDM** [reply]
I'm stuck creating the model. I'm trying to follow the page you pasted to copy the example, but I don't know where I should put all the class resistor, diode, etc. According to your response, I imagine it should go in another script, <your_run_script.py>. Is there something else I should put in this script so I can create the N2 diagram? After that, should I run the scripts in any order or just write the line you gave me?

**Reply by naylor-b** [reply]
To keep things simple, just copy all of the code shown in the example (including Resistor, Diode, etc) into the same python file and call it 'my_script.py' for example, and then run the command line `openmdao n2 my_script.py`

**Reply by robfalck** [reply]
There are a few different ways to create an N2 diagram.  Those that @JustinSGray listed work, but just calling `problem.run_model()` on a problem instance will, by default, generate an N2 diagram in a subdirectory named "reports".

Of course, this means that to get an N2 diagram from OpenMDAO you have to have built a model...or at least the skeleton of a model where various components have inputs and outputs that are interconnected - the compute methods don't have to be functional at this point.

If you're trying to demonstrate data flow between disciplines of your model, [XDSM diagrams](https://mdolab.engin.umich.edu/wiki/xdsm-overview) are also an option. The pyxdsm allows these to be developed and they're a bit more lightweight than N2 diagrams. Which you choose to use is just a matter of user preference and the situation, but I think XDSM diagrams are better for communicating the model functionality, while N2 diagrams are a more in-depth model debugging tool.

**Reply by Mr-PDM** [reply]
Thanks to everyone for your help, and I'm sorry  couldn't be more concise in my question. I think the XDSM diagram could fit more in this situation, but I will keep working in the N2 Diagram if it's necessary. <3

**Reply by Mr-PDM** [reply]
Speaking about XDSM, i find this error when i run the examples, I can't find what I am missing

Traceback (most recent call last):
  File "c:\Users\-\OneDrive\Escritorio\pyXDSM-main\examples\mdf.py", line 31, in <module>
    x.write("mdf")
  File "C:\Users\-\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0\LocalCache\local-packages\Python310\site-packages\pyxdsm\XDSM.py", line 703, in write
    subprocess.run(command, check=True)
  File "C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.10_3.10.3056.0_x64__qbz5n2kfra8p0\lib\subprocess.py", line 503, in run
    with Popen(*popenargs, **kwargs) as process:
  File "C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.10_3.10.3056.0_x64__qbz5n2kfra8p0\lib\subprocess.py", line 971, in __init__
    self._execute_child(args, executable, preexec_fn, close_fds,
  File "C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.10_3.10.3056.0_x64__qbz5n2kfra8p0\lib\subprocess.py", line 1456, in _execute_child
    hp, ht, pid, tid = _winapi.CreateProcess(executable, args,
FileNotFoundError: [WinError 2] El sistema no puede encontrar el archivo especificado

**Reply by JustinSGray** [reply]
you should post an bug report with a sample script to that repo. Supporting pyXDSM is outside the scope of what can be solved here.

**Comment by JustinSGray** [comment]
If you want an XDSM instead, you can check out the [pyXDSM repo](https://github.com/mdolab/pyXDSM). 

I actually highly encourage folks to build XDSMs using pyXDSM before starting to build an OpenMDAO model. Alternatively, if you find the N2 more useful then you can build a skeleton OpenMDAO model that just has components with I/O and connections. But no actual compute code.
